.circleci/config.yml

```diff
@@ -93,53 +117,195 @@ commands:
           command: ./scripts/changed-files.sh
 
 jobs:
-  build:
-    executor: build-executor
-    parallelism: 5
+  prefetch-crates:
+    executor: test-executor
+    description: Prefetch cargo crates for subsequent jobs.
     steps:
       - build_setup
       - run:
           name: Git Hooks and Checks
           command: ./scripts/git-checks.sh
       - run:
-          name: Linting
+          name: Fetch workspace dependencies over network
+          command: cargo fetch
+      - save_cargo_package_cache
+      - persist_to_workspace:
+          root: /home/circleci/project
+          paths:
+            - git
+            - registry
+            - .package-cache
+  lint:
+    executor: test-executor
+    description: Run Rust linting tools.
+    steps:
+      - build_setup
+      - run:
+          name: Git Hooks and Checks
+          command: ./scripts/git-checks.sh
+      - attach_workspace:
+          at: /home/circleci/project
+      - restore_cargo_package_cache
+      - run:
+          name: cargo lint
+          command: cargo x lint
+      - run:
+          name: cargo clippy
+          command: cargo xclippy --workspace --all-targets
+      - run:
+          name: cargo fmt
+          command: cargo xfmt --check
+      - run:
+          name: cargo guppy
           command: |
-            [[ $CIRCLE_NODE_INDEX =~ [1234] ]] || cargo x lint
-            [[ $CIRCLE_NODE_INDEX =~ [1234] ]] || cargo xclippy --workspace --all-targets
-            [[ $CIRCLE_NODE_INDEX =~ [1234] ]] || cargo xfmt --check
-            [[ $CIRCLE_NODE_INDEX =~ [1234] ]] || cargo install cargo-guppy --git http://github.com/calibra/cargo-guppy --rev 8b2bc45c0cd6323a7a2b8170ddad6d2a5b79047b
-            [[ $CIRCLE_NODE_INDEX =~ [1234] ]] || [[ -z $(cargo guppy dups --target x86_64-unknown-linux-gnu --kind directthirdparty) ]]
+            cargo install cargo-guppy \
+              --git http://github.com/calibra/cargo-guppy \
+              --rev 8b2bc45c0cd6323a7a2b8170ddad6d2a5b79047b
+            [[ -z $(cargo guppy dups --target x86_64-unknown-linux-gnu \
+              --kind directthirdparty) ]]
+  build-dev:
+    executor: build-executor
+    description: Development Build
+    steps:
+      - build_setup
+      - attach_workspace:
+          at: /home/circleci/project
+      - restore_cargo_package_cache
       - run:
-          name: Build Release
+          name: Biuld default
           command: |
-            [[ $CIRCLE_NODE_INDEX =~ [0234] ]] || RUST_BACKTRACE=1 cargo build -j 16 --release
+            RUST_BACKTRACE=1 cargo build -j 16
+      - run:
+          name: Build libra-swarm
+          command: RUST_BACKTRACE=1 cargo build -j 16 -p libra-swarm
+      - run:
+          name: Build cluster-test
+          command: RUST_BACKTRACE=1 cargo build -j 16 -p cluster-test
+      - run:
+          name: Build libra-fuzzer
+          command: RUST_BACKTRACE=1 cargo build -j 16 -p libra-fuzzer
```

Comment:
is there no support for variables? it would be nice to have this line RUST_BACKTRACE=1 .... -p to be called something like $RUST_BUILD or $LIBRA_BUILD

Reply:
Agreed, it's boiler plate :( 

There is distinction between the two.  The var used in the "name: " line is part of the yaml config, and it needs to be resolved when the job is created.  If we use env var defined in runtime, the job will appear as "Build $LIBRA_BUILD" in the job page.  The var used in the "command:" line needs to be resolved when the command is executed in the job. 

We can use a single var between the two if we want to. But that var needs to be configured as env var in Circle's admin config, not inside this config.  I'd prefer keep all things in one place if possible.

Reply:
I just removed the "name:" and used command as the step's name.  This should make it simpler.

Reply:
To follow up on @davidiw's question, CircleCI seems to not have any way to do arbitrary substitution unfortunately. You can call jobs with parameters, but other than env vars, it doesn't seem to have anything.